[tests] Add support for running EndToEnd tests on helix with the `aspire` workload

radical opened the pull request:
# Problem:

1. We want to run some EndToEnd tests on CI, which can `dotnet run` an aspire project,
and allow individual tests to interact with the services.
This requires:

	- Ability to build, and run an aspire project - IOW, a sdk with the `aspire` workload installed.
	- `docker`

2. Also, allow using `TestProject.*` in `tests/testproject`, in two modes:
    - `in-tree` test run which directly reference aspire projects, and repo targets
    - `out-of-tree` test runs which uses a SDK+workload based on local build output

# Solution:

### SDK+workload

- SDK with version from `global.json` is installed into `artifacts/bin/dotnet-{none,latest}`
    - The `aspire` workload is installed into the sdk using the nuget packages in `artifacts/packages` with a version like `8.0.0-{dev,ci}`
- this also allows running tests that do `dotnet new aspire-starter; dotnet run` same as a user
- this utilizes an internal nuget from `dotnet/runtime`

### TestProject

- This can switch between the two test run modes using the msbuild property `$(TestsRunningOutsideRepo)`
    - when running `in-tree` the test project directly references hosting targets, and aspire projects via `ProjectReference`
    - when running `out-of-tree` the `ProjectReferences` and imports are replaced with `PackageReferences` to the Aspire nugets
- Default is to run `in-tree`

### Helix

- The tests are built on the build machine
- The testproject, and the sdk+workload is sent to helix
  - where the tests run using `dotnet` from the sdk+workload path
- Since `docker` is needed to helix, this is enabled only for `Linux` in this PR. Blocked on https://github.com/dotnet/dnceng/issues/2067 for windows support.

## TL;dr or How do I use this?

### Steps to prepare for `out-of-tree` runs

1. `.\build.cmd -pack` - to build all the nugets
2. `dotnet build tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.csproj /t:InstallWorkloadUsingArtifacts /p:Configuration=<config>`
    - this will install the sdk, and the `aspire` workload using the nugets from `artifacts/packages`

- at this point the sdk is usable outside the repo too. Note: you need to add `artifacts/packages/$(Configuration)/Shipping` as a nuget source so the locally built packages can be picked up.

### Using it from VS

- No difference when running the EndToEnd tests in `in-tree` mode
- For switching to `out-of-tree` add `<TestsRunningOutsideRepo>true</TestsRunningOutsideRepo>` to `tests/Aspire.EndToEnd.Tests/Directory.props` *before* any imports.
    - tests cannot be run at this point as they will fail complaining about `artifacts/bin/dotnet-latest` being missing
    - Install the sdk+workload following the steps above
    - Run/debug the tests normally now, and they will be using the sdk
    - Also note that in this case the testproject is run from the bindir for `Aspire.EndToEnd.Tests`, so a path like `artifacts/bin/Aspire.EndToEnd.Tests/Debug/net8.0/testassets/testproject/`

### Using it from command line

- No difference when running the EndToEnd tests in `in-tree` mode
- Install the sdk+workload following the steps above
- When running the tests you can either:
    - set `<TestsRunningOutsideRepo>true</TestsRunningOutsideRepo>` to `tests/Aspire.EndToEnd.Tests/Directory.props` before any imports
    - or set the environment variable `TestsRunningOutsideRepo=true`

### Other details

- `tests/Shared/WorkloadTesting` - has sources essentially from `dotnet/runtime`'s `Wasm.Build.Tests`
- TestProject supports skipping individual resources like `--skip-resources oracledatabase,cosmos`
- `.runsettings` with 15min timeout
- `xunit.runner.json` - to get diagnostic messages
- The core infrastructure is not specific for `Aspire.EndToEnd.Tests`, and will get used by other test projects in the future

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2534)

radical committed:
testproject: Switch between Project and Package references

radical committed:
testproject: Add support for skipping resources

radical committed:
Aspire.EndToEnd.Tests: update

- add a xunit.runner.json, .runsettings
- update tests to skip resources as needed
- update test fixture to work with sdk+workload for testing

radical committed:
Workload testing

- reference workload.testing nuget from dotnet/runtime
- Extra some workload properties needed to a Workload.props, used in
  more than once place now

- Microsoft.NET.Sdk.Aspire: set `PackageType=''` to the manifest nuget is not
  `dotnetplatform`, and can be installed directly by the workload
  testing tasks.

radical committed:
Add common infrastructure code for workload testing

radical committed:
Add bits to support preparing tests for helix

radical committed:
remaining tidbits

radical committed:
CI: Build sdk+workload for sending to helix

radical committed:
testproject.AppHost: use Debug logging

radical committed:
cleanup

radical commented:
- [ ] Questions:
    - [ ] should a forced container cleanup be done at the end?
    - [ ] nuget-cache - Is it OK to expect the dev to manually delete aspire packages from the cache?
        - PR uses a new nuget-cache path for every run, which is slow on slow connections, but good on fast ones
    - [ ] Microsoft.NET.Sdk.Aspire: set `PackageType=''` to the manifest nuget is not`dotnetplatform`, and can be installed directly by the workload testing tasks. Check that nothing else is affected, and this is an acceptable change.

TODO:
    - [ ] nuget.helix.config - Use the root `nuget.config` and insert the built-nugets source

joperezr commented:
Talked offline with @radical, marking as ready for review as this is ready for feedback.

eerhardt commented on tests/Shared/WorkloadTesting/ProcessExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+
+internal static class ProcessExtensions
+{
+    // private static int RunProcessAndWaitForExit(string fileName, string arguments, TimeSpan timeout, out string stdout)
```
This should be deleted.

eerhardt commented on tests/Shared/WorkloadTesting/BuildEnvironment.cs:
```diff
@@ -0,0 +1,159 @@
+
+        HasSdkWithWorkload = expectSdkWithWorkload;
+        sdkForWorkloadPath = Path.GetFullPath(sdkForWorkloadPath);
+        DefaultBuildArgs = string.Empty;
```
Does this ever get set to anything other than `string.Empty`?

radical replied:
No, I can remove this. Though this would come in handy when adding build tests which will be new'ing up projects. Not needed right now though.

radical replied:
I tried removing it, and it triggers some other refactoring. Can we keep this in though, as it will be useful when we add more tests?

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs:
```diff
@@ -1,94 +1,111 @@
+            // that components wired up into this project have health checks enabled.
+            await _integrationServicesFixture.IntegrationServiceA.WaitForHealthyStatusAsync("http");
+        }
+        catch
+        {
+            _testOutput.WriteLine($"[{DateTime.Now}] <<<< FAILED VerifyHealthyOnIntegrationServiceA --");
+            await _integrationServicesFixture.DumpDockerInfoAsync();
+            throw;
```
Is it possible to do this in a Dispose() method on the class? That way it isn't duplicated across all the tests in the class.

radical replied:
Yes, but then it will run for passing tests too. If that's acceptable then I'll move it there.

eerhardt replied:
Maybe just refactor the code so it isn't duplicated in all the tests if we can't tell if the test failed or not in Dispose.

eerhardt commented on tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj:
```diff
@@ -2,10 +2,62 @@
   <PropertyGroup>
     <TargetFramework>$(NetCurrent)</TargetFramework>
+
+    <!-- no docker support on helix/windows yet -->
+    <RunTestsOnHelix Condition="'$(OS)' != 'Windows_NT'">true</RunTestsOnHelix>
+    <SkipTests Condition="'$(OS)' == 'Windows_NT'">true</SkipTests>
```
Can we just say `RunTestsOnHelix=true` all the time, and then do these checks inside the tests? Similar to how I did it in RabbitMQ in https://github.com/dotnet/aspire/pull/2345. It just feels cleaner that way.

radical replied:
Yes we can do it like that also. The reason for "also" is that if we don't have any tests that would get run, then not skipping the whole thing would mean:
1. installing sdk+workload - extra step on CI
2. sending this work item to helix with the big payload (sdk) for nothing to run, which dnceng kinda frowns upon.

I think we can add the attributes also, and for the specific case (windows) where nothing can run, we skip in the project file too. Thoughts?

eerhardt replied:
How long until we are able to run these tests on Windows?

eerhardt commented on tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj:
```diff
@@ -2,10 +2,62 @@
+
+    <BuiltNuGetsDir>$(ArtifactsShippingPackagesDir)</BuiltNuGetsDir>
+    <PackageVersionForWorkloadManifests>$(PackageVersion)</PackageVersionForWorkloadManifests>
+    <DefineConstants Condition="'$(TestsRunningOutOfTree)' == 'true'">TESTS_RUNNING_OUT_OF_TREE;$(DefineConstants)</DefineConstants>
```
Can we get this from an env var when the test runs so we don't have to have an `#if` in the code?

radical replied:
I had to add this just for the VS run, so you need to set only one property in one file to have the tests switch.
Using env var works perfectly well when running from command line, but I don't know what the workflow can be for setting it in case of VS. Open to suggestions.

eerhardt commented on eng/pipelines/templates/BuildAndTest.yml:
```diff
@@ -42,10 +43,22 @@ steps:
+                -restore
+                -p:ContinuousIntegrationBuild=true
+                -p:Configuration=${{ parameters.buildConfig }}
+                -t:InstallWorkloadUsingArtifacts
+                -bl:${{ parameters.repoLogPath }}/WorkloadInstallForTesting.binlog
+                $(Build.SourcesDirectory)/tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj
```
Is it weird to pick the EndToEnd tests to have this "InstallWorkloadUsingArtifacts" target? It seems like something that should be coming from an "infrastructure" project. Not one of the test projects.

radical replied:
Yes, I was thinking of fixing it in a follow up though. But I can try to do it here too.

eerhardt commented on tests/testproject/TestProject.AppHost/TestProgram.cs:
```diff
@@ -5,12 +5,29 @@
     private TestProgram(string[] args, Assembly assembly, bool includeIntegrationServices, bool includeNodeApp, bool disableDashboard)
     {
-        if (args.Contains("--disable-dashboard"))
+        IList<TestResourceNames> resourcesToSkip = [];
```
Do we actually need this `resourcesToSkip`? If we just skipped the cosmos and oracle tests, wouldn't that be enough? Yes it would still spin up those containers during the tests, but nothing would try to connect to them.

I don't really like all these if conditions everywhere. It really clutters up the code.

radical replied:
We have one test that health checks on the whole thing, and that would continue to fail.

> I don't really like all these if conditions everywhere. It really clutters up the code.
agree!

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:
```diff
@@ -2,40 +2,92 @@
+    private readonly IMessageSink _diagnosticMessageSink;
+    private readonly TestOutputWrapper _testOutput;
+
+    public IntegrationServicesFixture(IMessageSink messageSink)
+    {
+        _diagnosticMessageSink = messageSink;
```
(nit) ctor parameter names should match the field/property names.

radical commented on NuGet.config:
```diff
@@ -12,13 +12,22 @@
   </packageSources>
   <packageSourceMapping>
+    <!-- used for installing test workloads for resolving aspire workload manifest -->
+    <packageSource key="nuget-local">
```
https://github.com/dotnet/runtime/blob/f135699310bccfa83b3b441d2f4f1571fe8df480/src/tasks/WorkloadBuildTasks/InstallWorkloadFromArtifacts.cs#L262

I agree that it does seem odd that you can't directly see what this links up to in the repo. I can have the package source name be passed to the task, which would be in aspire repo.

joperezr replied:
Does this throw a warning at all on the initial restore before the task has a chance to replace the source? Did we also consider manually passing in `-s <path>` to a manual invocation of restore instead of having this in the config?

radical replied:
No warnings. Wouldn't `-s path` only add a package source but not the mapping, and because of that restoring from the source would fail. That is what I hit when I removed this mapping.

radical replied:
Once https://github.com/dotnet/runtime/pull/99277 is merged this entry can be removed, and it would be added at build time.

radical committed:
Use the version property in Directory.Packages.props

radical committed:
address review feedback from @ eerhardt

radical committed:
Rename TestsRunningOutOfTree -> TestsRunningOutsideOfRepo

radical committed:
Bump Microsoft.NET.Runtime.WorkloadTesting.Internal to latest(9.0.0-preview.3.24151.4)

radical committed:
address review feedback from @ eerhardt

radical committed:
Merge remote-tracking branch 'origin/main' into e2e-testing-helix

# Conflicts:
#	NuGet.config

radical committed:
fix merge

joperezr commented on eng/Version.Details.xml:
```diff
@@ -115,6 +115,10 @@
       <Sha>b4fa7f2e1e65ef49881be2ab2df27624280a8c55</Sha>
       <SourceBuild RepoName="source-build-reference-packages" ManagedOnly="true" />
     </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.WorkloadTesting.Internal" Version="9.0.0-preview.3.24151.4">
```
NIT: Can we move this new dependency up to the section where we have the rest of the dotnet-runtime repo dependencies?

joperezr commented on eng/Versions.props:
```diff
@@ -7,6 +7,7 @@
     <VersionPrefix>8.0.0</VersionPrefix>
     <PreReleaseVersionLabel>preview.5</PreReleaseVersionLabel>
     <UseVSTestRunner>true</UseVSTestRunner>
+    <MicrosoftNETRuntimeWorkloadTestingInternalVersion>9.0.0-preview.3.24151.4</MicrosoftNETRuntimeWorkloadTestingInternalVersion>
```
NIT: please move bellow with the rest of the package versions.

joperezr commented on eng/workloads/workloads.csproj:
```diff
@@ -10,8 +10,6 @@
     <WorkloadOutputPath>$(ArtifactsBinDir)workloads/</WorkloadOutputPath>
     <WorkloadOutputPath Condition="'$(workloadArtifactsPath)' != ''">$(workloadArtifactsPath)/</WorkloadOutputPath>
-    <PackageSource>$(ArtifactsShippingPackagesDir)</PackageSource>
-    <PackageSource Condition="'$(workloadPackagesPath)' != ''">$(workloadPackagesPath)/</PackageSource>
```
Does this mean that this cannot get overriden any longer? cc: @danegsta in case he is aware if we depend on this.

radical replied:
Where does `$(workloadPackagesPath)` get set? I didn't find any reference to it in the repo.

danegsta replied:
Should be fine as long as the workload still builds correctly

radical committed:
address review feedback from @ joperezr

joperezr commented on eng/dcppack/dcppack.csproj:
```diff
@@ -39,8 +39,16 @@
 
+  <Target Name="AddPackageFiles" BeforeTargets="_GetPackageFiles">
+    <ItemGroup>
+      <DcpFiles Include="$(NuGetPackageRoot)microsoft.developercontrolplane.$(DcpPlatform)/$(MicrosoftDeveloperControlPlanedarwinamd64PackageVersion)/tools/**/*" />
```
NIT: Can we use the GeneratePackagePath property instead? Or does that not work when using PackageDownload?

radical replied:
It doesn't seem to work for `PackageDownload`.

joperezr commented on eng/dashboardpack/dashboardpack.csproj:
```diff
@@ -36,7 +38,7 @@
 
     <MakeDir Directories="$(DashboardPublishedArtifactsOutputDir)/$(DashboardRuntime)" />
-    <ZipDirectory SourceDirectory="$(DotNetOutputPath)Aspire.Dashboard/$(Configuration)/$(TargetFramework)/$(DashboardRuntime)/publish" DestinationFile="$(DashboardPublishedArtifactsOutputDir)/$(DashboardRuntime)/aspire-dashboard-$(DashboardRuntime).zip" />
+    <ZipDirectory SourceDirectory="$(DotNetOutputPath)Aspire.Dashboard/$(Configuration)/$(TargetFramework)/$(DashboardRuntime)/publish" DestinationFile="$(DashboardPublishedArtifactsOutputDir)/$(DashboardRuntime)/aspire-dashboard-$(DashboardRuntime).zip" Overwrite="true" />
```
Why do we need to override?

radical replied:
We don't. I had added it for something else and forgot to remove!

radical replied:
Scratch that. This is useful for local builds when you might build the project even when the zip already exists.

joperezr commented on eng/dashboardpack/dashboardpack.csproj:
```diff
@@ -26,8 +26,10 @@
+  <Target Name="BeforeBuild" BeforeTargets="_GetPackageFiles">
+    <!-- separate restore step to allow targets from nugets to get imported in the publish step -->
+    <MSBuild Projects="../../src/Aspire.Dashboard/Aspire.Dashboard.csproj" Targets="restore" Properties="Configuration=$(Configuration);Platform=$(Platform);TargetFramework=$(TargetFramework);RuntimeIdentifier=$(DashboardRuntime)" />
+    <MSBuild Projects="../../src/Aspire.Dashboard/Aspire.Dashboard.csproj" Targets="publish" Properties="Configuration=$(Configuration);Platform=$(Platform);TargetFramework=$(TargetFramework);RuntimeIdentifier=$(DashboardRuntime);_Some_unique_prop=$([System.Guid]::NewGuid())" />
```
Can you explain why are we passing in this extra property? This is breaking incremental build. Essentially Restore is happening again given there are new global properties.

danegsta replied:
Is this solving a particular issue for the helix test scenario? Aspire.Dashboard should already be restored as part of the main build before this project runs.

radical replied:
This was a stale change which I had to add earlier because a restored nuget package was adding targets, but the project was not re-evaluated thus the targets were not available.
Subsequently, I had changed the way we install the workload for testing, but missed removing this change. Thank you for questioning it!

radical replied:
This wasn't stale, instead the conditions were reproducible only on CI. 
TL;dr - this was needed because it needed fresh restore, and a project re-evaluation to load targets from `grpc.tools` nuget.

Details:

On CI we used `./build.sh` for the main build, and then `.dotnet.sh build` to install the workload, which hit the target above.
- `build.sh` - sets `NUGET_PACKAGES` to `$repoRoot/.packages`
- and `./dotnet.sh build` defaulted to `$HOME/.nuget/packages`

So,
1. `./build.sh` - restores nugets to `$repoRoot/.packages`, and imports targets from the grpc.tools nuget.
2. `.dotnet build` - runs the above line and runs `publish` target
    a. which imports the `nuget.g.{props,targets}` for the E2E project, but it fails to import the `grpc.tools` targets because it is now looking at the wrong path to the restored package!
    b. Even if the `restore` target is run before this `publish`, the msbuild project is not re-evaluated and thus the `gprc.tools` targets don't get imported
Instead, if we `restore`, and then `publish` but with a new property added then it triggers a re-evaluation of the project, which causes the `grpc.tools` targets to get imported, and run

Now that I have figured out the reason for all this mess - `./build.sh vs ./dotnet.sh build`, I have:
1. changed to use `./build.sh` for both the cases
2. And have the target instead be triggered based on a property passed to the command on CI only.

This fixes the issue correctly.

joperezr commented on src/Microsoft.NET.Sdk.Aspire/Microsoft.NET.Sdk.Aspire.csproj:
```diff
@@ -5,6 +5,9 @@
     <Description>.NET Aspire workload manifest</Description>
+
+    <!-- this allows installing the manifest as a nuget, which is useful for testing -->
+    <PackageType />
```
Just Curious, does this change anything in the nuspec?

radical replied:
Yes, this removes:
```xml
    <packageTypes>
      <packageType name="DotnetPlatform" />
    </packageTypes>
```
.. from the `nuspec`. I checked other manifest nugets and they do have this specified, but I skipped this for the mono/wasm manifests, and it hasn't been an issue.

In a follow up for `dotnet/runtime`, I can change the way the nuget is installed which would remove the need for `PackageType` to be `''`.

joperezr replied:
mmm, I suppose this property is set via Arcade. It would be good to check what the implications of that are with someone from coreeng team.

joperezr commented on src/Shared/Workload.targets:
```diff
@@ -12,13 +14,7 @@
     <DotNetOutputPath>$(RepoRoot)artifacts/bin/</DotNetOutputPath>
     <DotNetDirectory>$(DotNetOutputPath)dotnet/</DotNetDirectory>
     <DotNetPacksDirectory>$(DotNetDirectory)packs/</DotNetPacksDirectory>
-    <VersionBand Condition=" '$(VersionBand)' == '' ">$([System.Text.RegularExpressions.Regex]::Match($(MicrosoftDotnetSdkInternalPackageVersion), `^\d+\.\d+\.\d`))00</VersionBand>
```
Just curious, why extract only those properties and not the rest?

radical replied:
The extracted properties define the version, and band used for the workload. The remaining properties are relevant to the various projects that import it for building packages. Some of these common properties could still be moved to the props file, but not all.

radical committed:
Move more workload testing bits into Aspire.Workload.Testing.*

- Splitting the props, and targets was required before the order of
  imports changed
- Patch root nuget.config to add the relevant package source for use
  with workload testing.

radical committed:
dashboardpack.csproj: Remove a stale change which is not required now as

.. the way we run the install workload targets was changed.

radical committed:
Add clarifying comment

radical committed:
fix build

radical committed:
Merge remote-tracking branch 'origin/main' into e2e-testing-helix

radical committed:
dashboardpack.csproj - use just publish again

radical committed:
Use build.sh instead of dotnet.sh to run InstallWorkloadUsingArtifacts target so the same NUGET_PACKAGES gets used

radical committed:
fix build

radical committed:
ensure helix.props gets patched before zip is generated for the tests

radical committed:
Merge remote-tracking branch 'origin/main' into e2e-testing-helix

radical committed:
Set TEST_LOG_PATH so binlog from the testproject can be accessed

radical commented on tests/nuget.helix.config:
```diff
@@ -0,0 +1,61 @@
+    </packageSource>
+    <packageSource key="orleans-nightly">
+      <package pattern="Microsoft.Orleans.*" />
+    </packageSource>
```
`nuget.helix.config` is generated from the root `nuget.config` now.

radical committed:
Merge remote-tracking branch 'origin/main' into e2e-testing-helix

# Conflicts:
#	tests/testproject/TestProject.AppHost/TestProgram.cs

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:
```diff
@@ -2,40 +2,92 @@
     {
-        var appHostDirectory = Path.Combine(GetRepoRoot(), "tests", "testproject", "TestProject.AppHost");
+        var appHostDirectory = Path.Combine(BuildEnvironment.TestProjectPath, "TestProject.AppHost");
+        if (TestsRunningOutsideOfRepo)
+        {
+            _testOutput.WriteLine("");
+            _testOutput.WriteLine($"****************************************");
+            _testOutput.WriteLine($"   Running tests outside-of-repo");
+            _testOutput.WriteLine($"   TestProject: {appHostDirectory}");
+            _testOutput.WriteLine($"   Using dotnet: {BuildEnvironment.DotNet}");
+            _testOutput.WriteLine($"****************************************");
+            _testOutput.WriteLine("");
+        }
+
+        {
+            using var cmd = new DotNetCommand(BuildEnvironment, _testOutput, label: "build")
+                .WithWorkingDirectory(appHostDirectory);
+
+            (await cmd.ExecuteAsync(CancellationToken.None, $"build -bl:{Path.Combine(BuildEnvironment.LogRootPath, "testproject-build.binlog")} -v m"))
+                .EnsureSuccessful();
+        }
```
This is super confusing formatting. When you skim the code it looks like the `using var cmd =....` code only runs in the `else` of the `if (TestsRunningOutsideOfRepo)` condition.

Instead of this custom scoping, we should just extract lines 70-76 to a method. It will make this code much easier to read.

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:
```diff
@@ -2,40 +2,92 @@
         _appHostProcess = new Process();
-        _appHostProcess.StartInfo = new ProcessStartInfo("dotnet", "run -- --disable-dashboard")
+
+        string processArguments = $"run --no-build -- ";
```
Why did we split build from run? Was it so we could split the binlog?

radical replied:
Yes

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:
```diff
@@ -2,40 +2,92 @@
+            using var cmd = new DotNetCommand(BuildEnvironment, _testOutput, label: "build")
+                .WithWorkingDirectory(appHostDirectory);
+
+            (await cmd.ExecuteAsync(CancellationToken.None, $"build -bl:{Path.Combine(BuildEnvironment.LogRootPath, "testproject-build.binlog")} -v m"))
```
Why `-v m`?

radical replied:
Because most of the time the build is going to pass and doesn't give any useful info for issues in tests. If the build fails then those messages will show up, and a `testproject-build.binlog` will be available for debugging offline.

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:
```diff
@@ -151,15 +287,30 @@ public async Task DisposeAsync()
-        while (directory != null && !Directory.Exists(Path.Combine(directory, ".git")))
+    private static IList<string> GetResourcesToSkip()
+    {
+        List<string> resourcesToSkip = new();
```
```suggestion
        HashSet<string> resourcesToSkip = new();
```

? We don't really want dupes in this collection.

radical committed:
address review feedback from @ eerhardt

radical committed:
CI: Try adding back oracledb

radical committed:
address review feedback from @ eerhardt

radical committed:
Merge remote-tracking branch 'origin/main' into e2e-testing-helix

radical committed:
address review feedback from @ eerhardt

radical committed:
Revert "CI: Try adding back oracledb"

This reverts commit 5be48e01c6e221d9be19b990b7b97c68bbf12a58.

radical commented:
@eerhardt I tried to add oracledb back, but it failed because the test timeout was 5minutes. I think we continue with that in a follow up PR, so reverting that change.

joperezr commented on eng/dashboardpack/dashboardpack.csproj:
```diff
@@ -26,7 +26,7 @@
   <Target Name="Build" />
 
-  <Target Name="BeforeBuild" BeforeTargets="Build">
+  <Target Name="BeforeBuild" BeforeTargets="_GetPackageFiles">
```
Why do we need this change? In general, it's best to avoid depending on "private" tragets (Targets that start with an underscore)

radical replied:
I wanted to limit this target to only when we are generating a nuget.

Is this target needed only for generating the zip file? Is the zip file used for the dashboard docker image?

joperezr replied:
This target mainly accomplishes two things:

- Adds the right content to the each of the dashboard packs (which actually may no longer be needed now that we have the dashboard as a tool) cc: @danegsta 
- Zips up the published output as that will be uploaded as a blob so that it can then be used to build the docker image for the dashboard.

radical replied:
Reverted.

joperezr commented on eng/dcppack/dcppack.csproj:
```diff
@@ -39,8 +39,16 @@
 
   <Target Name="Build" />
 
+  <Target Name="AddPackageFiles" BeforeTargets="_GetPackageFiles">
```
Same comment here, if we could instead hook up to one of the "public" targets: https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#advanced-extension-points-to-create-customized-package

joperezr reviewed (approved):
Added a few comments/suggestions but the rest LGTM.

radical committed:
dcppack.csproj: Use intended targets for adding package files

https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#targetsfortfmspecificcontentinpackage

addresses review feedback from @ joperezr

radical committed:
dashboardpack.csproj: avoid depending on private targets, addresses review feedback from @ joperezr

radical committed:
Aspire.csproj: Use the PackageType='' workaround only for CI when running tests

radical committed:
Merge remote-tracking branch 'origin/main' into e2e-testing-helix

radical commented on eng/pipelines/templates/BuildAndTest.yml:
```diff
@@ -21,6 +21,7 @@ parameters:
 steps:
   - script: ${{ parameters.buildScript }}
             -restore -build
+            -pack
```
Good point. This pack could conflict with the subsequent pack, since that one is signed. I will investigate this in a separate PR. And it will really be needed only when we have this running on windows again.
For now, I will add `-pack` only for linux, where the separate `pack` step is not defined at all.

radical committed:
Use -pack with build step only on linux as it is needed only for e2e testing

radical committed:
[ci] helix: clean up docker containers before, and after a test run

radical committed:
[tests] increase e2e tests timeout from 5mins to 10mins

radical committed:
Disable E2E tests for Pomelo : https://github.com/dotnet/aspire/issues/2690

radical commented:
Follow-up work is being tracked in https://github.com/dotnet/aspire/issues/2687 .

radical committed:
Skip pomelo resource, so VerifyHealthyOnIntegrationServiceA doesn't get stuck waiting for i t

radical committed:
Merge remote-tracking branch 'origin/main' into e2e-testing-helix

# Conflicts:
#	tests/testproject/TestProject.IntegrationServiceA/Program.cs

radical committed:
fix build

radical committed:
fix timeout typo

radical committed:
Correctly disable pomelo
